ntoskrnl/config/ntapi.c

```diff
@@ -1937,20 +1939,37 @@ NtUnloadKey2(IN POBJECT_ATTRIBUTES TargetKey,
     }
 
     /* Call the internal API. Note that CmUnloadKey() unlocks the registry only on success. */
-    Status = CmUnloadKey(KeyBody->KeyControlBlock, Flags);
+    Status = CmUnloadKey(KeyBody->KeyControlBlock, Flags, LockExHeld);
 
     /* Check if we failed, but really need to succeed */
-    if ((Status == STATUS_CANNOT_DELETE) && (Flags == REG_FORCE_UNLOAD))
-    {
-        /* TODO: We should perform another attempt here */
-        _SEH2_TRY
-        {
-            DPRINT1("NtUnloadKey2(%wZ): We want to force-unload the hive but couldn't unload it: Retrying is UNIMPLEMENTED!\n", TargetKey->ObjectName);
-        }
-        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    if ((Status == STATUS_CANNOT_DELETE) && (Flags != REG_FORCE_UNLOAD))
+    {
+        /*
+         * We failed to unload the hive normally, this usually happens
+         * when we could not safely remove empty cache entries.
+         */
+        CmpReleaseTwoKcbLockByKey(ChildConv, ParentConv);
+        ExReleasePushLockExclusive(&CmpLoadHiveLock);
+        CmpUnlockRegistry();
+
+        /* Lock the entire registry down, at the cost of performance */
+        CmpLockRegistryExclusive();
```

Comment:
Honestly, this function is not a hot path and IMO all this stuff with individual locks is useless "optimization" at the cost of increased code complexity. It should be simplified to always lock the entire registry and good.